java-frontend/src/main/java/org/sonar/java/model/GeneratedFile.java

```diff
@@ -80,21 +81,26 @@ public void addSmap(SmapFile smap) {
 
     final Map<Integer, Location> lines = new HashMap<>();
 
-    private SourceMapImpl() {
+    private SourceMapImpl(Function<Path, Optional<InputFile>> fileResolver) {
       for (SmapFile sm : smapFiles) {
         for (SmapFile.LineInfo lineInfo : sm.getLineSection()) {
-          for (int i = 0; i < lineInfo.repeatCount; i++) {
-            int inputLine = lineInfo.inputStartLine + i;
-            Path inputFile = sm.getUriRoot().resolve(sm.getFileSection().get(lineInfo.lineFileId).sourcePath);
-            LocationImpl location = new LocationImpl(inputFile, inputLine, inputLine);
-            int outputStart = lineInfo.outputStartLine + (i * lineInfo.outputLineIncrement);
-            int outputEnd = lineInfo.outputStartLine + ((i + 1) * lineInfo.outputLineIncrement) - 1;
-            // when outputLineIncrement == 0, end will be less than start (looks like bug in spec)
-            outputEnd = max(outputStart, outputEnd);
-            for (int j = outputStart; j <= outputEnd; j++) {
-              lines.merge(j, location, LocationImpl::mergeLocations);
-            }
-          }
+          Path sourcePath = sm.getUriRoot().resolve(sm.getFileSection().get(lineInfo.lineFileId).sourcePath);
+          Optional<InputFile> inputFile = fileResolver.apply(sourcePath);
```

Comment:
We can maybe avoid looking up the same lineFileId for every line by having just a check of the difference of lineInfo.lineFileId with the previous one as in most cases (if not all) we should not have a difference.

Reply:
in fact I had a look into Jasper code and it seems that 1 JSP always maps to single Java file, so this was unnecessary